### PR TITLE
Finished the first working version of the API.

### PR DESCRIPTION
**Pull request by radiantdinosaurs**
Created functions that check base64 file type (making sure it is JPEG or PNG), decode the file, and save to disk. Replaced basic image recognition (default image in classify_image.py) for image recognition of user supplied base64 encoded image. Added in proper error responses for unsupported file types and undefined body.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,46 +1,138 @@
...
+app.use(bodyParser.urlencoded({limit: "50mb", extended: true, parameterLimit: 50000}));
+
+/**
+ * Writes a base64 string for a JPEG file to disk as a JPEG file
+ * @param {string} base64String - base64 string of JPEG file
+ */
+function writeJpegToDisk(base64String) {
+    let base64Data = base64String.replace(/^data:image\/jpeg;base64,/, ""); // stripping for decoding
+    // writing file to disk as "out.jpeg"
+    fs.writeFile("out.jpeg", base64Data, "base64", function (error) {
```
Comment (andrewvora) [inline app.js]: This wouldn't handle parallel server requests. If two requests came in, it's possible you're operating on the wrong "out.jpeg". I recommend using the JS equivalent of Java's `UUID.randomUUID()`

File: app.js [hunk 2]
```diff
@@ -1,46 +1,138 @@
...
+        if (error !== null) {
+            console.log(error);
+        }
```
Comment (andrewvora) [inline app.js]: You should probably handle the error somehow. Just letting it silently fail is asking for trouble

File: app.js [hunk 3]
```diff
@@ -1,46 +1,138 @@
...
+    });
```
Comment (andrewvora) [inline app.js]: Have you considered using the arrow syntax instead of an anonymous function?

File: app.js [hunk 4]
```diff
@@ -1,46 +1,138 @@
...
+ */
+function writePngToDisk(base64String) {
+    const buffer = new Buffer(base64String.split(/,\s*/)[1], "base64");
+    // using middleware to change image from PNG to JPEG
+    pngToJpeg({quality: 90})(buffer).then(function (output) {
+        // writing file to disk as "out.jpeg"
+        fs.writeFile("out.jpeg", output, function (error) {
+            if (error !== null) {
+                console.log(error);
+            }
```
Comment (andrewvora) [inline app.js]: Same point here about the silent failure

File: app.js [hunk 5]
```diff
@@ -1,46 +1,138 @@
...
+/**
+ * Writes a base64 string for a PNG file to disk as a JPEG file
+ * @param {string} base64String - base64 string of PNG file
+ */
+function writePngToDisk(base64String) {
+    const buffer = new Buffer(base64String.split(/,\s*/)[1], "base64");
+    // using middleware to change image from PNG to JPEG
+    pngToJpeg({quality: 90})(buffer).then(function (output) {
+        // writing file to disk as "out.jpeg"
+        fs.writeFile("out.jpeg", output, function (error) {
```
Comment (andrewvora) [inline app.js]: Same comment about unique filenames

File: app.js [hunk 6]
```diff
@@ -1,46 +1,138 @@
...
+ * @returns {string} fileType - assignment of file type, which will either be PNG, JPEG, or undefined
+ */
+function checkFileType(base64String) {
+    let fileType = base64String.substring(0, 15); // getting the data/image type label
+    if (fileType.includes("png")) {
+        return "PNG"; // image is a PNG
+    } else if (fileType.includes("jpeg")) {
+        return "JPEG"; // image is a JPEG
+    } else {
+        return "undefined"; // image is not PNG or JPEG, so set as "undefined," because it cannot be used
```
Comment (andrewvora) [inline app.js]: I think you can just return `undefined` instead of the string, `"undefined"`

File: app.js [hunk 7]
```diff
@@ -1,46 +1,138 @@
...
+}
+
+/**
+ * Confirms the file type of a base64 string, whether PNG, JPEG, or other
+ * @param {string} base64String - base64 string of file
+ * @returns {string} fileType - assignment of file type, which will either be PNG, JPEG, or undefined
+ */
+function checkFileType(base64String) {
+    let fileType = base64String.substring(0, 15); // getting the data/image type label
+    if (fileType.includes("png")) {
```
Comment (andrewvora) [inline app.js]: This throws an exception if fileType is undefined. This function should be self-contained for the most part. If the logic is copied anywhere, someone won't think that this logic isn't enforcing non-null parameters

File: app.js [hunk 8]
```diff
@@ -1,46 +1,138 @@
...
 
 /**
  * Parses the output of the TensorFlow trained model script (classify_image.py) to a JSON object
  * @param {string} result - Output from classify_image.py
  */
 function parseTensorFlowResult(result) {
-    let resultLines = result.toString().split("\n");
-    let resultObject = {"image_labels": []};
+    let sloth = false; // true if image contains a sloth
+    let resultLines = result.toString().split("\n"); // splitting up TensorFlow's result by newline
```
Comment (andrewvora) [inline app.js]: Consider adding handling for where result is undefined

File: app.js [hunk 9]
```diff
@@ -1,46 +1,138 @@
...
+            resultObject.image_labels.push({
+                "name": fields[0].trim(), // remove trailing whitespaces
+                "score": fields[1].replace(/[^\d.]/g, "") // remove all except numbers and decimal points
             });
+            // check if the line contains a label for sloth with confidence higher than 70%
+            if(fields[0].includes("sloth") && !fields[0].includes("Ursus ursinus")) {
+                if(fields[1].replace(/[^\d.]/g, "") > 0.70) {
+                    sloth = true;
+                }
+            }
```
Comment (andrewvora) [inline app.js]: You should definitely have unit tests for this logic. There may be edge cases you're not handling

Comment (andrewvora) [inline app.js]: Consider using arrow syntax here too for the callback passed to `forEach`

File: app.js [hunk 10]
```diff
@@ -1,46 +1,138 @@
...
+    let fileType = checkFileType(base64String); // checking the file type of the base64
+    if (fileType.includes("PNG")) {
+        writePngToDisk(base64String);
+    } else if (fileType.includes("JPEG")) {
+        writeJpegToDisk(base64String);
+    } else {
+        return null; // return null if the file type isn't PNG or JPEG
+    }
+    return "../slothbucket/out.jpeg"; // return file path if the file type is PNG or JPEG
+}
```
Comment (andrewvora) [inline app.js]: Same point here about `fileType` being `undefined`

File: app.js [hunk 11]
```diff
@@ -1,46 +1,138 @@
...
+function processImage(base64String) {
+    let fileType = checkFileType(base64String); // checking the file type of the base64
+    if (fileType.includes("PNG")) {
+        writePngToDisk(base64String);
+    } else if (fileType.includes("JPEG")) {
+        writeJpegToDisk(base64String);
+    } else {
+        return null; // return null if the file type isn't PNG or JPEG
+    }
+    return "../slothbucket/out.jpeg"; // return file path if the file type is PNG or JPEG
```
Comment (andrewvora) [inline app.js]: Potential danger with `out.jpeg` being non-unique

File: app.js [hunk 12]
```diff
@@ -1,46 +1,138 @@
...
-// HTTP get test path (temporary path for manual testing)
-app.get("/test", function (req, res) {
-    child_process.execFile("python", ["classify_image.py"], {
-        cwd: "../root"
-    }, function (error, result) {
-        parseTensorFlowResult(result);
-        res.send(result.toString().split("\n"));
-    });
+// HTTP POST for any TensorFlow image recognition with bonus sloth check
+app.post("/sloth-check", function (req, res) {
```
Comment (andrewvora) [inline app.js]: Consider using arrow-syntax

File: app.js [hunk 13]
```diff
@@ -1,46 +1,138 @@
...
+            });
+        } else {
+            // if filePath is null, that means the image wasn't able to be processed (e.g., it wasn't a JPEG or PNG)
+            res.status(500).send({status:500, message:"File type not supported"});
+        }
+    }
+    else {
+        // body was undefined or null
+        res.status(500).send({status:500, message:"Cannot read undefined body"});
+    }
```
Comment (andrewvora) [inline app.js]: This logic as well as the associated functions might be better fit being put into a module that gets required in `app.js`

**Review by andrewvora — CHANGES_REQUESTED** [review]
Left some comments regarding code style and structure

**Review by andrewvora — CHANGES_REQUESTED** [review]
You may want to consider putting the app.js file into an `app/` folder. Then update `package.json` to have `start` run `node app/app.js`

**Review by andrewvora — APPROVED** [review]
The logic is a mess and it's hard to read, but given that you're refactoring this into "modules" that can be addressed at a later date. I'm approving this based on the fact the happy path works

Some quick TODOs for a future PR:

* break error handling into their own functions or controllers
* avoid long inline conditionals, i.e. if ( A && B && C && D && E). Prefer `const isF = A && B && C && D && E`.
* Your null and undefined checks don't work the way you think they do

**Comment by andrewvora** [comment]
Follow these steps. I leave finding out the commands up to you. Make sure you learn about unfamiliar commands and understand each step. Don't blindly follow steps.

1. Make sure the development branch is up-to-date.
2. Check out your feature branch.
3. Rebase the development branch ONTO your feature branch.
4. Take care of any merge conflicts.
5. Finish the rebase.
6. Push up your changes. You will need to force a push since you will be altering the git history.

Alternatively, you can just do steps 1-2, then merge development ONTO your feature branch. 

Read about the differences between merging and rebasing: https://www.atlassian.com/git/tutorials/merging-vs-rebasing

I prefer rebasing.
